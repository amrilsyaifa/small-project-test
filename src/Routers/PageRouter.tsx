import React from 'react';
import { Router, useLocation, useRouter } from 'wouter';

export interface IPageRouter {
  base: string;
  children?: React.ReactNode;
}

const PageRouter: React.FC<IPageRouter> = ({ base, children }) => {
  const router = useRouter();
  const [location] = useLocation();
  const nestedBase = `${router.base}${base}`;
  const parentLocation = `${router.base}${location}`;

  if (!parentLocation.startsWith(nestedBase)) return null;
  return (
    <Router base={nestedBase} key={nestedBase}>
      {children}
    </Router>
  );
};

export default PageRouter;
