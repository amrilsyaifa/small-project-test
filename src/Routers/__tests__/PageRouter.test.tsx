import { render } from '@testing-library/react';
import React from 'react';
import MockSeparator from '../../Reusables/Mocks/ParentChildren.Mock';
import { useLocation, useRouter } from 'wouter';
import PageRouter from '../PageRouter';

jest.mock('wouter', () => ({
  __esModule: true,
  Router: (props) => <MockSeparator mockLabel="AppRouter" {...props} />,
  useLocation: jest.fn(),
  useRouter: jest.fn()
}));

const mockUseLocation = useLocation as jest.Mock;
const mockUseRouter = useRouter as jest.Mock;

describe('PageRouter', () => {
  describe('should render if in correct url', () => {
    it('root level', () => {
      mockUseLocation.mockReturnValue(['root/correct/anything']);
      mockUseRouter.mockReturnValue({ base: '' });
      const { container } = render(
        <PageRouter base="root/correct">This should display</PageRouter>
      );
      expect(container).toMatchSnapshot();
    });
    it('children level', () => {
      mockUseLocation.mockReturnValue(['/anything']);
      mockUseRouter.mockReturnValue({ base: '/root/correct' });
      const { container } = render(<PageRouter base="/anything">This should display</PageRouter>);
      expect(container).toMatchSnapshot();
    });
  });

  describe('should NOT render if not in correct url', () => {
    it('root level', () => {
      mockUseLocation.mockReturnValue(['root/correct/anything']);
      mockUseRouter.mockReturnValue({ base: '' });
      const { container } = render(
        <PageRouter base="not-correct">This should NOT display</PageRouter>
      );
      expect(container).toMatchSnapshot();
    });
    it('children level', () => {
      mockUseLocation.mockReturnValue(['/anything']);
      mockUseRouter.mockReturnValue({ base: '/root/correct' });
      const { container } = render(<PageRouter base="not-correct">This should display</PageRouter>);
      expect(container).toMatchSnapshot();
    });
  });
});
