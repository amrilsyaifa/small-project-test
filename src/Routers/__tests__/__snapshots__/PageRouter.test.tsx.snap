// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageRouter should NOT render if not in correct url children level 1`] = `<div />`;

exports[`PageRouter should NOT render if not in correct url root level 1`] = `<div />`;

exports[`PageRouter should render if in correct url children level 1`] = `
<div>
  <div
    data-props="{\\"base\\":\\"/root/correct/anything\\"}"
    data-testid="AppRouter"
  >
    This should display
  </div>
</div>
`;

exports[`PageRouter should render if in correct url root level 1`] = `
<div>
  <div
    data-props="{\\"base\\":\\"root/correct\\"}"
    data-testid="AppRouter"
  >
    This should display
  </div>
</div>
`;
